// thirdweb/dist/cjs/extensions/erc1155/__generated__/DropERC1155/write/freezeBatchBaseURI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isFreezeBatchBaseURISupported = isFreezeBatchBaseURISupported;
exports.encodeFreezeBatchBaseURIParams = encodeFreezeBatchBaseURIParams;
exports.encodeFreezeBatchBaseURI = encodeFreezeBatchBaseURI;
exports.freezeBatchBaseURI = freezeBatchBaseURI;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa07ced9e";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_index",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `freezeBatchBaseURI` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `freezeBatchBaseURI` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isFreezeBatchBaseURISupported } from "thirdweb/extensions/erc1155";
 *
 * const supported = await isFreezeBatchBaseURISupported(contract);
 * ```
 */
async function isFreezeBatchBaseURISupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "freezeBatchBaseURI" function.
 * @param options - The options for the freezeBatchBaseURI function.
 * @returns The encoded ABI parameters.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeFreezeBatchBaseURIParams } "thirdweb/extensions/erc1155";
 * const result = encodeFreezeBatchBaseURIParams({
 *  index: ...,
 * });
 * ```
 */
function encodeFreezeBatchBaseURIParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.index]);
}
/**
 * Encodes the "freezeBatchBaseURI" function into a Hex string with its parameters.
 * @param options - The options for the freezeBatchBaseURI function.
 * @returns The encoded hexadecimal string.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeFreezeBatchBaseURI } "thirdweb/extensions/erc1155";
 * const result = encodeFreezeBatchBaseURI({
 *  index: ...,
 * });
 * ```
 */
function encodeFreezeBatchBaseURI(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeFreezeBatchBaseURIParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "freezeBatchBaseURI" function on the contract.
 * @param options - The options for the "freezeBatchBaseURI" function.
 * @returns A prepared transaction object.
 * @extension ERC1155
 * @example
 * ```ts
 * import { freezeBatchBaseURI } from "thirdweb/extensions/erc1155";
 *
 * const transaction = freezeBatchBaseURI({
 *  contract,
 *  index: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function freezeBatchBaseURI(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.index];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
