// thirdweb/dist/cjs/react/native/ui/connect/ConnectedButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectedButton = ConnectedButton;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_native_1 = require("react-native");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const wallet_js_1 = require("../../../core/utils/wallet.js");
const index_js_1 = require("../../design-system/index.js");
const useActiveWalletChain_js_1 = require("../../hooks/wallets/useActiveWalletChain.js");
const Skeleton_js_1 = require("../components/Skeleton.js");
const WalletImage_js_1 = require("../components/WalletImage.js");
const button_js_1 = require("../components/button.js");
const text_js_1 = require("../components/text.js");
function ConnectedButton(props) {
    const theme = (0, CustomThemeProvider_js_1.parseTheme)(props.theme);
    const { account, wallet } = props;
    const walletChain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const { ensAvatarQuery, addressOrENS, balanceQuery } = (0, wallet_js_1.useConnectedWalletDetails)(props.client, walletChain, account, props.detailsButton?.displayBalanceToken);
    return ((0, jsx_runtime_1.jsx)(button_js_1.ThemedButton, { theme: theme, onPress: () => {
            props.openModal();
        }, style: {
            paddingHorizontal: index_js_1.spacing.md,
            paddingVertical: index_js_1.spacing.smd,
        }, children: (0, jsx_runtime_1.jsxs)(react_native_1.View, { style: styles.row, children: [(0, jsx_runtime_1.jsx)(WalletImage_js_1.WalletImage, { theme: theme, size: 40, wallet: wallet, ensAvatar: ensAvatarQuery.data }), (0, jsx_runtime_1.jsxs)(react_native_1.View, { style: styles.col, children: [(0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, type: "defaultSemiBold", style: {
                                color: theme.colors.primaryButtonText,
                            }, children: addressOrENS }), balanceQuery.data ? ((0, jsx_runtime_1.jsxs)(text_js_1.ThemedText, { theme: theme, type: "subtext", style: {
                                fontSize: index_js_1.fontSize.sm,
                            }, children: [Number(balanceQuery.data.displayValue).toFixed(3), " ", balanceQuery.data?.symbol] })) : ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { theme: theme, style: { width: 80, height: 16 }, color: theme.colors.secondaryText }))] })] }) }));
}
const styles = react_native_1.StyleSheet.create({
    row: {
        flexDirection: "row",
        flex: 1,
        gap: index_js_1.spacing.md,
        alignItems: "center",
    },
    col: {
        flexDirection: "column",
        gap: index_js_1.spacing.xxs,
    },
});
