// thirdweb/dist/cjs/react/native/ui/connect/InAppWalletUI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InAppWalletUI = InAppWalletUI;
exports.OtpLogin = OtpLogin;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const react_native_1 = require("react-native");
const nativeStorage_js_1 = require("../../../../utils/storage/nativeStorage.js");
const index_js_1 = require("../../../../wallets/in-app/native/auth/index.js");
const storage_js_1 = require("../../../core/utils/storage.js");
const index_js_2 = require("../../design-system/index.js");
const RNImage_js_1 = require("../components/RNImage.js");
const button_js_1 = require("../components/button.js");
const input_js_1 = require("../components/input.js");
const spacer_js_1 = require("../components/spacer.js");
const spinner_js_1 = require("../components/spinner.js");
const text_js_1 = require("../components/text.js");
const svgs_js_1 = require("../icons/svgs.js");
const defaultAuthOptions = [
    "email",
    "phone",
    "google",
    "facebook",
    "apple",
];
const socialIcons = {
    google: svgs_js_1.GOOGLE_ICON,
    facebook: svgs_js_1.FACEBOOK_ICON,
    apple: svgs_js_1.APPLE_ICON,
};
function InAppWalletUI(props) {
    const { wallet, theme } = props;
    const config = wallet.getConfig();
    const authOptions = config?.auth?.options || defaultAuthOptions;
    const socialLogins = authOptions.filter((x) => x === "google" || x === "apple" || x === "facebook");
    const [inputMode, setInputMode] = (0, react_1.useState)("email");
    return ((0, jsx_runtime_1.jsxs)(react_native_1.View, { style: styles.container, children: [(0, jsx_runtime_1.jsx)(react_native_1.View, { style: styles.row, children: socialLogins.map((auth) => ((0, jsx_runtime_1.jsx)(SocialLogin, { auth: auth, ...props }, auth))) }), inputMode === "email" ? ((0, jsx_runtime_1.jsx)(PreOtpLogin, { auth: "email", ...props })) : ((0, jsx_runtime_1.jsx)(button_js_1.ThemedButtonWithIcon, { theme: theme, title: "Email address", icon: svgs_js_1.EMAIL_ICON, onPress: () => setInputMode("email") })), inputMode === "phone" ? ((0, jsx_runtime_1.jsx)(PreOtpLogin, { auth: "phone", ...props })) : ((0, jsx_runtime_1.jsx)(button_js_1.ThemedButtonWithIcon, { theme: theme, title: "Phone number", icon: svgs_js_1.PHONE_ICON, onPress: () => setInputMode("phone") }))] }));
}
function SocialLogin(props) {
    const { theme, wallet, auth, client, connectMutation } = props;
    // TODO (rn) - move this onPress and state up
    const strategy = props.auth;
    const [selectedAuth, setSelectedAuth] = (0, react_1.useState)();
    const connectInAppWallet = (0, react_1.useCallback)(() => {
        setSelectedAuth(strategy);
        connectMutation.connect(async () => {
            await wallet.connect({
                client,
                strategy: auth,
            });
            await (0, storage_js_1.setLastAuthProvider)(auth, nativeStorage_js_1.nativeLocalStorage);
            return wallet;
        });
    }, [connectMutation, auth, client, wallet, strategy]);
    return ((0, jsx_runtime_1.jsx)(react_native_1.View, { style: [
            styles.socialIconContainer,
            { borderColor: theme.colors.borderColor },
        ], children: connectMutation.isConnecting && selectedAuth === auth ? ((0, jsx_runtime_1.jsx)(spinner_js_1.ThemedSpinner, { color: theme.colors.primaryText })) : ((0, jsx_runtime_1.jsx)(react_native_1.TouchableOpacity, { onPress: connectInAppWallet, disabled: connectMutation.isConnecting, children: (0, jsx_runtime_1.jsx)(RNImage_js_1.RNImage, { theme: theme, size: 38, data: socialIcons[auth] }) }, strategy)) }));
}
function PreOtpLogin(props) {
    const { theme, auth, client, setScreen, wallet } = props;
    const [phoneOrEmail, setPhoneNumberOrEmail] = (0, react_1.useState)("");
    const sendCode = (0, react_query_1.useMutation)({
        mutationFn: async (options) => {
            const { auth, phoneOrEmail } = options;
            if (auth === "phone") {
                await (0, index_js_1.preAuthenticate)({
                    client,
                    strategy: auth,
                    phoneNumber: phoneOrEmail,
                });
            }
            else {
                await (0, index_js_1.preAuthenticate)({
                    client,
                    strategy: auth,
                    email: phoneOrEmail,
                });
            }
        },
    });
    return ((0, jsx_runtime_1.jsx)(input_js_1.ThemedInputWithSubmit, { theme: theme, placeholder: auth === "phone" ? "Phone number" : "Email address", onChangeText: setPhoneNumberOrEmail, value: phoneOrEmail, keyboardType: auth === "phone" ? "phone-pad" : "email-address", onSubmit: () => sendCode.mutate({
            auth,
            phoneOrEmail,
        }, {
            onSuccess: (_, vars) => {
                if (vars.auth === "phone") {
                    setScreen({
                        screen: "otp",
                        auth: {
                            strategy: vars.auth,
                            phoneNumber: vars.phoneOrEmail,
                        },
                        wallet,
                    });
                }
                else {
                    setScreen({
                        screen: "otp",
                        auth: { strategy: vars.auth, email: vars.phoneOrEmail },
                        wallet,
                    });
                }
            },
            onError: (error) => {
                // TODO (rn) - handle error toast or input error border/label
                react_native_1.Alert.alert("Error", error.message);
            },
        }), isSubmitting: sendCode.isPending }));
}
function OtpLogin(props) {
    const { theme, auth, wallet, client, connectMutation } = props;
    const [verificationCode, setVerificationCode] = (0, react_1.useState)("");
    const connectInAppWallet = async () => {
        if (!verificationCode || !verificationCode)
            return;
        await connectMutation.connect(async () => {
            if (auth.strategy === "phone") {
                await wallet.connect({
                    client,
                    strategy: auth.strategy,
                    phoneNumber: auth.phoneNumber,
                    verificationCode,
                });
            }
            else {
                await wallet.connect({
                    client,
                    strategy: auth.strategy,
                    email: auth.email,
                    verificationCode,
                });
            }
            await (0, storage_js_1.setLastAuthProvider)(auth.strategy, nativeStorage_js_1.nativeLocalStorage);
            return wallet;
        });
    };
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsxs)(react_native_1.View, { style: {
                    flexDirection: "column",
                    justifyContent: "center",
                    alignItems: "center",
                }, children: [(0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, style: { color: theme.colors.secondaryText }, children: "Enter the verification code sent to" }), (0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, type: "defaultSemiBold", children: auth.strategy === "phone" ? auth.phoneNumber : auth.email })] }), (0, jsx_runtime_1.jsx)(spacer_js_1.Spacer, { size: "xs" }), (0, jsx_runtime_1.jsx)(input_js_1.ThemedInput, { theme: theme, placeholder: "Verification code", onChangeText: setVerificationCode, value: verificationCode, keyboardType: "number-pad" }), (0, jsx_runtime_1.jsx)(button_js_1.ThemedButton, { theme: theme, onPress: connectInAppWallet, variant: "accent", disabled: connectMutation.isConnecting, children: connectMutation.isConnecting ? ((0, jsx_runtime_1.jsx)(spinner_js_1.ThemedSpinner, { color: theme.colors.accentButtonText })) : ((0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, type: "defaultSemiBold", style: { color: theme.colors.accentButtonText }, children: "Verify" })) }), connectMutation.error && ((0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, type: "subtext", style: { color: theme.colors.danger }, children: connectMutation.error.message }))] }));
}
const styles = react_native_1.StyleSheet.create({
    container: {
        flexDirection: "column",
        gap: index_js_2.spacing.md,
    },
    row: {
        flexDirection: "row",
        flexWrap: "wrap",
        gap: index_js_2.spacing.md,
        justifyContent: "space-evenly",
    },
    socialIconContainer: {
        flex: 1,
        flexDirection: "column",
        alignItems: "center",
        justifyContent: "center",
        borderStyle: "solid",
        borderWidth: 1,
        borderRadius: index_js_2.radius.lg,
        height: 60,
    },
});
